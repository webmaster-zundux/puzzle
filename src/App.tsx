import "@fontsource-variable/open-sans";
import type { FC } from "react";
import { memo } from "react";
import "./styles.css";

import { PuzzleDataProvider } from "./components/PuzzleDataProvider";
import { PuzzlePage } from "./pages/PuzzlePage";

export const App: FC = memo(function App() {
  return (
    <PuzzleDataProvider>
      <PuzzlePage />
    </PuzzleDataProvider>
  );
});
