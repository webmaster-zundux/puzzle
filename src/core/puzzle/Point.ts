import type { Id } from "./BaseEntity";
import { BaseEntity } from "./BaseEntity";
import type { PositionDelta } from "./PositionDelta";

export const generatePointIdFromPosition = (x = 0, y = 0): Id => {
  return `${x}:${y}`;
};

export type PositionString = Id;

export const generatePositionString = (x = 0, y = 0): PositionString => {
  return `${x}:${y}`;
};

interface PointCreation {
  x: number;
  y: number;
  id?: string;
}

export class Point extends BaseEntity {
  constructor(
    public x: number,
    public y: number,
    id?: string,
  ) {
    const pointId = id ? id : generatePointIdFromPosition(x, y);
    super({ id: pointId });
  }

  static create({ x, y, id }: PointCreation) {
    return new this(x, y, id);
  }

  static clone(point: Point): Point {
    return Point.create({ x: point.x, y: point.y });
  }

  static getPointAtZero() {
    return new Point(0, 0);
  }

  static isPointInAreaWithCenterPointAndAreaSideSize(point: Point, areaCenterPoint: Point, areaSideSize = 0): boolean {
    const cx = areaCenterPoint.x;
    const cy = areaCenterPoint.y;
    const areaHalfSideSize = areaSideSize / 2;

    const tx = point.x;
    const ty = point.y;

    const p0 = new Point(cx - areaHalfSideSize, cy - areaHalfSideSize);
    const p2 = new Point(cx + areaHalfSideSize, cy + areaHalfSideSize);

    if (tx < p0.x || tx > p2.x) {
      return false;
    }

    if (ty < p0.y || ty > p2.y) {
      return false;
    }

    return true;
  }

  static isPointInAreaByCornerPoints(point: Point, areaPointP0: Point, areaPointP2: Point): boolean {
    const tx = point.x;
    const ty = point.y;

    const p0 = areaPointP0;
    const p2 = areaPointP2;

    if (tx < p0.x || tx > p2.x) {
      return false;
    }

    if (ty < p0.y || ty > p2.y) {
      return false;
    }

    return true;
  }

  static addDelta(point: Point, positionDelta: PositionDelta): Point {
    return new Point(point.x + positionDelta.dx, point.y + positionDelta.dy);
  }
}
