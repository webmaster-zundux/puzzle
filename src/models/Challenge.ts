import type { Id } from "../core/puzzle/BaseEntity";
import { uuid } from "../utils/uuid";

export type ChallengeId = Id;

const generateChallengeIdFunction = () => {
  const newUUID = uuid();
  return newUUID;
};

const EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT =
  import.meta.env.VITE_TESTING_CHALLENGE_ID_GENERATION_EXPOSE_TO_WINDOW_OBJECT === "1";
if (EXPOSE_CHALLENGE_ID_GENERATION_TO_WINDOW_OBJECT) {
  // @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
  if (typeof window?.generateChallengeId !== "function") {
    Object.defineProperty(window, "generateChallengeId", {
      value: generateChallengeIdFunction,
      writable: false,
      configurable: false,
    });
  }
}

// @ts-expect-error property 'generateChallengeId' does not exist on type 'Window & typeof globalThis'
export const generateChallengeId = window?.generateChallengeId ?? generateChallengeIdFunction;
