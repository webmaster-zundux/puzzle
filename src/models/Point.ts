export interface IPoint {
  x: number;
  y: number;

  id: string;
}

export interface IPointCreation {
  x?: number;
  y?: number;

  id?: string | number;
}

export class Point {
  static instanceIndex = 0;

  constructor(
    public x: number = 0,
    public y: number = 0,
    public id?: string | number,
  ) {
    let newId: string;
    if (typeof id === "string") {
      newId = id;
    } else {
      if (id) {
        newId = Point.createId(id);
      } else {
        newId = Point.createIdBasedOnInstanceIndex();
      }
    }
    this.id = newId;
  }

  static createId(id: number | string): string {
    const className = this.name;
    return `${className}-${id}`;
  }
  static createIdBasedOnInstanceIndex(): string {
    const className = this.name;
    return `${className}-${Point.instanceIndex++}`;
  }

  static create({ x, y }: IPointCreation): Point {
    return new this(x, y);
  }

  static clone(args: Point): Point {
    return this.create(args);
  }

  static getZeroPoint() {
    return new Point(0, 0);
  }

  add(point: Point) {
    const x = this.x + point.x;
    const y = this.y + point.y;

    return new Point(x, y);
  }

  subtract(point: Point) {
    const x = this.x - point.x;
    const y = this.y - point.y;

    return new Point(x, y);
  }

  cloneAndLimitToIntegerByMathFloorFunction() {
    const integerX = Math.floor(this.x);
    const integerY = Math.floor(this.y);

    return new Point(integerX, integerY);
  }

  toString() {
    const className = this.constructor.name;
    return `${className} {x: ${this.x}, y: ${this.y}}`;
  }
}
