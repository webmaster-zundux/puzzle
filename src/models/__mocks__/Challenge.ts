export const generateChallengeId = () => {
  const newUUID = "364bf6d0-6686-4892-be7d-618ac2fcdcc7";
  return newUUID;
};
