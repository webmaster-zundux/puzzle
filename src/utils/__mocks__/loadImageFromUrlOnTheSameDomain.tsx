import type * as originalModule from "../loadImageFromUrlOnTheSameDomain";

export const loadImageFromUrlOnTheSameDomain: typeof originalModule.loadImageFromUrlOnTheSameDomain = () => {
  return Promise.reject(undefined);
};
