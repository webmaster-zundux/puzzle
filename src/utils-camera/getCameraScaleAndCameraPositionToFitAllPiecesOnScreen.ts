import type { CacheableByCameraScaleCacheableRenderablePuzzle } from "../models/CacheableByCameraScaleCacheableRenderablePuzzle";
import { Point } from "../models/Point";
import { CANVAS_SCALE_MAX, CANVAS_SCALE_MIN, limitScaleByRange } from "./getNextZoomValue";

const FIT_TO_SCREEN_PADDING_TO_PUZZLE_HORIZONTAL_IN_PX = 50;
const FIT_TO_SCREEN_PADDING_TO_PUZZLE_VERTICAL_IN_PX = 50;

export const getCameraScaleAndCameraPositionToFitAllPiecesOnScreen = ({
  canvasWidth,
  canvasHeight,
  puzzle,
}: {
  canvasWidth: number;
  canvasHeight: number;
  puzzle: CacheableByCameraScaleCacheableRenderablePuzzle;
}): {
  scale: number;
  position: Point;
} => {
  const availableCanvasWidth = canvasWidth - FIT_TO_SCREEN_PADDING_TO_PUZZLE_HORIZONTAL_IN_PX * 2;

  const availableCanvasHeight = canvasHeight - FIT_TO_SCREEN_PADDING_TO_PUZZLE_VERTICAL_IN_PX * 2;

  const { p0, p2 } = puzzle.getBoundaryCornerPointsOfPiecesSpreadArea();
  const puzzleAreaTopLeftCorner = new Point(p0.x, p0.y);
  const puzzleAreaBottomRightCorner = new Point(p2.x, p2.y);

  const puzzleAreaWidth = puzzleAreaBottomRightCorner.x - puzzleAreaTopLeftCorner.x;
  const puzzleAreaHeight = puzzleAreaBottomRightCorner.y - puzzleAreaTopLeftCorner.y;

  let fitToScreenCoefficient = availableCanvasWidth / puzzleAreaWidth;
  const puzzleAreaHeightAfterScale = puzzleAreaHeight * fitToScreenCoefficient;
  if (puzzleAreaHeightAfterScale > availableCanvasHeight) {
    fitToScreenCoefficient = availableCanvasHeight / puzzleAreaHeight;
  }

  let newCameraScale = fitToScreenCoefficient;
  newCameraScale = limitScaleByRange(newCameraScale, CANVAS_SCALE_MIN, CANVAS_SCALE_MAX);

  const newAvailableCanvasWidth = (canvasWidth - FIT_TO_SCREEN_PADDING_TO_PUZZLE_HORIZONTAL_IN_PX * 2) / newCameraScale;

  const newAvailableCanvasHeight = (canvasHeight - FIT_TO_SCREEN_PADDING_TO_PUZZLE_VERTICAL_IN_PX * 2) / newCameraScale;

  const worldPaddingToPuzzleHorizontal = FIT_TO_SCREEN_PADDING_TO_PUZZLE_HORIZONTAL_IN_PX / newCameraScale;
  const worldPaddingToPuzzleVertical = FIT_TO_SCREEN_PADDING_TO_PUZZLE_VERTICAL_IN_PX / newCameraScale;

  const x =
    worldPaddingToPuzzleHorizontal + (newAvailableCanvasWidth - puzzleAreaWidth) / 2 - puzzleAreaTopLeftCorner.x;

  const y =
    worldPaddingToPuzzleVertical + (newAvailableCanvasHeight - puzzleAreaHeight) / 2 - puzzleAreaTopLeftCorner.y;

  const newCameraPosition = new Point(x, y);

  return {
    scale: newCameraScale,
    position: newCameraPosition,
  };
};
