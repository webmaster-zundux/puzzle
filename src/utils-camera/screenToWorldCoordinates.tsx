import { Point } from "../models/Point";

export const screenToWorldCoordinates = (
  point: Point,
  scale: number,
  worldOriginPoint: Point = Point.getZeroPoint(),
): Point => {
  const xInWorldScale = point.x / scale;
  const yInWorldScale = point.y / scale;

  const x = -worldOriginPoint.x + xInWorldScale;
  const y = -worldOriginPoint.y + yInWorldScale;

  return new Point(x, y);
};

export const worldToScreenCoordinates = (
  point: Point,
  scale: number,
  worldOriginPoint: Point = Point.getZeroPoint(),
): Point => {
  const x = (point.x + worldOriginPoint.x) * scale;
  const y = (point.y + worldOriginPoint.y) * scale;

  return new Point(x, y);
};

export const screenToWorldSize = (size: number, scale: number): number => {
  return size / scale;
};

export const worldToScreenSize = (size: number, scale: number): number => {
  return size * scale;
};
