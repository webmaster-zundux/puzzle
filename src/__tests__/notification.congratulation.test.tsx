import { screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { App } from "../App";
import * as Challenge from "../models/Challenge";
import { Point } from "../models/Point";
import * as getCameraScaleAndCameraPositionToFitAllPiecesOnScreenModule from "../utils-camera/getCameraScaleAndCameraPositionToFitAllPiecesOnScreen";
import * as screenToWorldCoordinatesModule from "../utils-camera/screenToWorldCoordinates";
import * as loadImageFromUrlOnTheSameDomainModule from "../utils/loadImageFromUrlOnTheSameDomain";
import { renderIntoDocumentBody } from "./__tests-utils__/renderIntoDocumentBody";
import {
  dragAndDropPuzzlePiece,
  openPuzzleCreationDialogByClickOnChangeNumberOfPiecesButtonInSidebar,
  openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar,
} from "./modals/modal-helpers";

const isCongratulationNotificationDisplaying = async ({ isShortCheck } = { isShortCheck: true }) => {
  await screen.findByRole("alert", { name: /puzzle was solved/i });

  const note = screen.getByRole("note", { name: /puzzle was solved/i });
  expect(note).toBeInTheDocument();
  if (isShortCheck) {
    return note;
  }

  const alert = screen.getByRole("alert", { name: /puzzle was solved/i });
  expect(alert).toHaveTextContent(/Congratulation! Puzzle was solved/i);

  expect(within(note).getByRole("button", { name: /close/i })).toBeInTheDocument();

  return note;
};

const mockSavedPuzzleState = () => {
  const LOCAL_STORAGE_KEY_NAME_PUZZLE_INFORMATION = "puzzleInformation";
  return vi.spyOn(localStorage, "getItem").mockImplementation((key: string) => {
    if (key === LOCAL_STORAGE_KEY_NAME_PUZZLE_INFORMATION) {
      return JSON.stringify({
        name: "puzzle name",
        challengeId: "364bf6d0-6686-4892-be7d-618ac2fcdcc7",
        numberOfPiecesPerWidth: 2,
        numberOfPiecesPerHeight: 1,
        pieceWidth: 50,
        connectionActivationAreaSideSizeFractionFromPieceSideSize: 0.2,
        imageSrc: "/demo-images/demo-image-1920x1080.png",
        imageOriginalSize: { width: 1920, height: 1080 },
        boundaryPoints: {
          tl: { x: 0, y: 0 },
          br: { x: 1914, y: 960 },
        },
        piecesPositions: [
          { id: "piece_0:0", position: { x: 0, y: 0 } },
          { id: "piece_1:0", position: { x: 200, y: 0 } },
        ],
        timeSpent: 0,
        pieceSideShapeName: "UNDERGROUND_RIVER_CIRCLE_BOTTOM_IN_ERODED_HILL",
        isSidebarOpen: true,
        canvasBackgroundColor: "dark-slate-gray",
      });
    }

    return null;
  });
};

const mockLoadedImageInstance = () => {
  vi.spyOn(loadImageFromUrlOnTheSameDomainModule, "loadImageFromUrlOnTheSameDomain").mockImplementation((imageSrc) => {
    console.log("mock loadImageFromUrlOnTheSameDomain", { imageSrc });

    let image;
    if (imageSrc === "/demo-images/demo-image-1920x1080.png") {
      image = new Image(1920, 1080);
    } else {
      image = new Image(250, 250);
    }
    image.src = imageSrc;
    return Promise.resolve(image);
  });
};

const mockCameraPosition = () => {
  vi.spyOn(screenToWorldCoordinatesModule, "screenToWorldCoordinates").mockImplementation(
    (point: Point, scale: number, worldOriginPoint: Point) => {
      console.log("mock screenToWorldCoordinates - params: ", {
        params: { point: point, scale, worldOriginPoint },
        "returns point as is": { x: point.x, y: point.y },
      });
      return point;
    },
  );
};

const mockGetCameraScaleAndCameraPositionToFitAllPiecesOnScreen = () => {
  return vi
    .spyOn(
      getCameraScaleAndCameraPositionToFitAllPiecesOnScreenModule,
      "getCameraScaleAndCameraPositionToFitAllPiecesOnScreen",
    )
    .mockImplementation(() => {
      console.log("mock getCameraScaleAndCameraPositionToFitAllPiecesOnScreen");
      return {
        scale: 1,
        position: new Point(0, 0),
      };
    });
};

const solvePuzzle = async () => {
  const puzzleView = screen.getByRole<HTMLCanvasElement>("application", { name: /puzzle view/i });
  expect(puzzleView).toBeInTheDocument();
  console.log("test-helper call - solvePuzzle", { boundingClientRect: puzzleView.getBoundingClientRect() });

  await dragAndDropPuzzlePiece(puzzleView, { clientX: 200, clientY: 0 }, { clientX: 50, clientY: 0 });
};

const initPuzzle = async () => {
  mockSavedPuzzleState();
  mockLoadedImageInstance();
  mockCameraPosition();
  const spyFitAllPieceOnScreen = mockGetCameraScaleAndCameraPositionToFitAllPiecesOnScreen();
  const { asFragment } = renderIntoDocumentBody(<App />);

  await waitFor(() => {
    expect(spyFitAllPieceOnScreen).toHaveBeenCalledOnce();
  });

  return {
    asFragment,
  };
};

describe(`"congratulation" notification`, () => {
  describe.sequential(`changes it's visibility state`, async () => {
    it("displays when a puzzle becomes solved", async () => {
      const { asFragment } = await initPuzzle();
      const renderWithoutDialog = asFragment();

      await solvePuzzle();

      await isCongratulationNotificationDisplaying({ isShortCheck: false });
      const sidebar = screen.getByRole("complementary", { name: /sidebar/i });
      expect(within(sidebar).getByRole("note", { name: /challenge progress/i })).toHaveTextContent(/progress:100%/i);

      expect(renderWithoutDialog).toMatchDiffSnapshot(
        asFragment(),
        { aAnnotation: "without notification", bAnnotation: "with notification" },
        `"congratulation" notification`,
      );
    });

    it(`hides when clicking on "close" button`, async () => {
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");
      const { asFragment } = await initPuzzle();
      const renderWithoutAlert = asFragment();
      expect(Challenge.generateChallengeId).not.toHaveBeenCalled();
      const initialChallengeSeed = screen.getByRole("note", { name: /challenge seed:/i }).textContent;
      expect(initialChallengeSeed).toBe("364bf6d0-6686-4892-be7d-618ac2fcdcc7");
      expect(screen.queryByRole("alert")).not.toBeInTheDocument();
      await solvePuzzle();
      const alert = await isCongratulationNotificationDisplaying();

      await userEvent.click(within(alert).getByRole("button", { name: /close/i }));

      const sidebar = screen.getByRole("complementary", { name: /sidebar/i });
      expect(within(sidebar).getByRole("note", { name: /challenge progress/i })).toHaveTextContent(/progress:100%/i);
      expect(screen.queryByRole("alert")).not.toBeInTheDocument();
      expect(renderWithoutAlert).toMatchDiffSnapshot(asFragment(), undefined, `without notification`);
      expect(Challenge.generateChallengeId).not.toHaveBeenCalledOnce();
      const challengeSeedAfterCanceledRestartOfPuzzle = screen.getByRole("note", {
        name: /challenge seed:/i,
      }).textContent;
      expect(challengeSeedAfterCanceledRestartOfPuzzle).toBe(initialChallengeSeed);
    });

    it("hides when a puzzle restarts", async () => {
      const user = userEvent;
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");
      const { asFragment } = await initPuzzle();
      const renderWithoutDialog = asFragment();
      const initialChallengeSeed = screen.getByRole("note", { name: /challenge seed:/i }).textContent;
      expect(Challenge.generateChallengeId).not.toHaveBeenCalled();
      expect(initialChallengeSeed).toBe("364bf6d0-6686-4892-be7d-618ac2fcdcc7");
      await solvePuzzle();
      await isCongratulationNotificationDisplaying();

      const dialog = await openSolvePuzzleAgainDialogByClickOnSolvePuzzleAgainButtonInSidebar({ user });
      await user.click(within(dialog).getByRole("button", { name: /solve again/i }));

      expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
      expect(Challenge.generateChallengeId).toHaveBeenCalledOnce();
      const secondChallengeSeed = screen.getByRole("note", { name: /challenge seed:/i }).textContent;
      expect(secondChallengeSeed).not.toBe(initialChallengeSeed);
      expect(secondChallengeSeed).toBe("b47663e4-95ab-4472-9344-0f165a6a7ff9");
      const sidebar = screen.getByRole("complementary", { name: /sidebar/i });
      expect(within(sidebar).getByRole("note", { name: /challenge progress/i })).toHaveTextContent(/progress:0%/i);

      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without notification`);
    });

    it.skip("hides when a new puzzle creates", async () => {
      const user = userEvent;
      vi.spyOn(Challenge, "generateChallengeId")
        .mockImplementationOnce(() => "b47663e4-95ab-4472-9344-0f165a6a7ff9")
        .mockImplementationOnce(() => "89e80c0c-186a-46db-b242-b147c88a2a3e");
      const { asFragment } = await initPuzzle();
      const renderWithoutDialog = asFragment();
      await solvePuzzle();
      await isCongratulationNotificationDisplaying();

      const dialog = await openPuzzleCreationDialogByClickOnChangeNumberOfPiecesButtonInSidebar({ user });
      await user.click(within(dialog).getByRole("button", { name: /create puzzle/i }));

      const sidebar = screen.getByRole("complementary", { name: /sidebar/i });
      expect(within(sidebar).getByRole("note", { name: /challenge progress/i })).toHaveTextContent(/progress:0%/i);

      expect(renderWithoutDialog).toMatchDiffSnapshot(asFragment(), undefined, `without notification`);
    });
  });
});
