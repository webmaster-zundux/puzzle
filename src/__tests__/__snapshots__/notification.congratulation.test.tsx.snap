// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`src/__tests__/notification.congratulation.test.tsx > "congratulation" notification > changes it's visibility state > displays when a puzzle becomes solved: "congratulation" notification 1`] = `
"Snapshot Diff:
- without notification
+ with notification

@@ -14,10 +14,53 @@
            class="_Canvas_72d137"
            id="canvas"
            role="application"
          />
          <div
+           aria-label="puzzle was solved"
+           class="_Notification_668b1d"
+           role="note"
+         >
+           <div
+             aria-label="puzzle was solved"
+             class="_Container_af0fbb"
+             role="alert"
+           >
+             <div
+               class="_Header_af0fbb"
+             >
+               Congratulation!
+             </div>
+             <div
+               class="_Spacer_af0fbb"
+             >
+               Â 
+             </div>
+             <div
+               class="_Message_af0fbb"
+             >
+               Puzzle was solved
+             </div>
+           </div>
+           <div
+             class="_CloseIcon_bea9a0"
+           >
+             <button
+               aria-label="close"
+               class="_ActionButton_b70b4c _NoBackground_b70b4c _NoBorder_b70b4c"
+               type="button"
+             >
+               <div
+                 class="_Icon_14488b _DoubleSize_14488b"
+                 data-src="/assets/fa-icons/xmark.svg"
+               >
+                 inline-svg-icon-mock
+               </div>
+             </button>
+           </div>
+         </div>
+         <div
            class="_FloatingInfoLabel_72d137"
          >
            <div
              class="_ZoomLabel_aa5be1"
            >
@@ -174,11 +217,11 @@
                class="_Label_59689d"
              >
                Time:
              </div>
              <time
-               class="_PassedTime_59689d _TimerIsPaused_59689d"
+               class="_PassedTime_59689d"
                datetime="PT0S"
                title="elapsed time"
              >
                0:00
              </time>
@@ -195,22 +238,21 @@
                  class="_Label_ae1a35"
                >
                  Progress:
                </div>
                <div
-                 class="_Value_ae1a35 _ZeroValue_ae1a35"
+                 class="_Value_ae1a35"
                >
-                 0%
+                 100%
                </div>
              </div>
              <div
                class="_ActionButtons_ae1a35"
              >
                <button
                  aria-label="restart puzzle"
                  class="_ActionButton_b70b4c"
-                 disabled=""
                  title="restart puzzle"
                  type="button"
                >
                  <div
                    class="_Icon_14488b""
`;

exports[`src/__tests__/notification.congratulation.test.tsx > "congratulation" notification > changes it's visibility state > hides when a puzzle restarts: without notification 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -447,11 +447,11 @@
            </div>
            <div
              aria-labelledby="challenge-seed-label"
              role="note"
            >
-             364bf6d0-6686-4892-be7d-618ac2fcdcc7
+             b47663e4-95ab-4472-9344-0f165a6a7ff9
            </div>
          </div>
        </aside>
      </main>
    </div>"
`;

exports[`src/__tests__/notification.congratulation.test.tsx > "congratulation" notification > changes it's visibility state > hides when clicking on "close" button: without notification 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -174,11 +174,11 @@
                class="_Label_59689d"
              >
                Time:
              </div>
              <time
-               class="_PassedTime_59689d _TimerIsPaused_59689d"
+               class="_PassedTime_59689d"
                datetime="PT0S"
                title="elapsed time"
              >
                0:00
              </time>
@@ -195,22 +195,21 @@
                  class="_Label_ae1a35"
                >
                  Progress:
                </div>
                <div
-                 class="_Value_ae1a35 _ZeroValue_ae1a35"
+                 class="_Value_ae1a35"
                >
-                 0%
+                 100%
                </div>
              </div>
              <div
                class="_ActionButtons_ae1a35"
              >
                <button
                  aria-label="restart puzzle"
                  class="_ActionButton_b70b4c"
-                 disabled=""
                  title="restart puzzle"
                  type="button"
                >
                  <div
                    class="_Icon_14488b""
`;
